File: MATLAB/SimpleInversion.m
classdef SimpleInversion < matlab.mixin.SetGetExactNames & matlab.mixin.CustomDisplay
    % Simple Inversion class
    properties (Constant)
        object_type = "inversion";
        inversion_type = "simple";
    end
    properties
        phi(:,:) {mustBeNumeric}
        defl_ax1(:,:) {mustBeNumeric}
        defl_ax2(:,:) {mustBeNumeric}
        
        pradformat_version string = pradformat_version()
        dr double
        source_object_dist double
        object_image_dist double
        spec_name string
        spec_mass double
        spec_charge double
        spec_energy double
        dr_ax2 double
        source_radius double
        label string
        description string
        file_date string = datestr(datetime('now'), 'yyyy-mm-dd')
        radiograph_filename string       
        fields_filename string       
    end
    properties (Hidden)
        % Categorize the above public properties as required or optional
        req_ds = []; % Required datasets
        opt_ds = ["phi", "defl_ax1", "defl_ax2"]; % Optional datasets
        req_atts = [...  % Required attributes
            "object_type", ...
            "inversion_type", ...
            "pradformat_version", ...
            "dr", ...
            "source_object_dist", ...
            "object_image_dist", ...
            "spec_name", ...
            "spec_mass", ...
            "spec_charge", ...
            "spec_energy", ...
            ];
        opt_atts = [...  % Optional attributes
            "dr_ax2", ...
            "source_radius", ...
            "label", ...
            "description", ...
            "file_date", ...
            "radiograph_filename", ...
            "fields_filename", ...
            ];
        % List for a second time any above properties that are constants
        % of the class (and are not to be overwritten when reading from file)
        const_atts = ["object_type", "inversion_type"]
    end
    methods (Access = protected)
       function propgrps = getPropertyGroups(obj)
          % Customized pretty printing of object
          % propgrp1 = matlab.mixin.util.PropertyGroup(cellstr(obj.req_ds), 'Datasets (Required)');
          propgrp2 = matlab.mixin.util.PropertyGroup(cellstr(obj.opt_ds), 'Datasets (Optional)');
          propgrp3 = matlab.mixin.util.PropertyGroup(cellstr(obj.req_atts), 'Attributes (Required)');
          propgrp4 = matlab.mixin.util.PropertyGroup(cellstr(obj.opt_atts), 'Attributes (Optional)');
          propgrps = [propgrp2 propgrp3 propgrp4];
       end
       function validate(obj)
            % Validate that all required properties have been set.
            % If some are not, throw an error.
            for i=1:length(obj.req_ds)
                ds = char(obj.req_ds(i));
                if isempty(get(obj, ds))
                    error(['Please assign a value to all required properties.' newline 'The following required dataset property is not yet assigned: "' char(ds) '".' newline 'Assign a value via "object.' char(ds) ' = value" and try again.'])
                end
            end           
            for i=1:length(obj.req_atts)
                att = char(obj.req_atts(i));
                if isempty(get(obj, att))
                    error(['Please assign a value to all required properties.' newline 'The following required attribute property is not yet assigned: "' char(att) '". ' newline 'Assign a value via "object.' char(att) ' = value" and try again.'])
                end
            end
       end
    end
    methods
        function obj = SimpleInversion(h5filename)
            % Constructor method - load object from file
            if nargin == 1
                % Load in the prad object from h5filename
                load_prad_group(obj, h5filename, '/')
            end
        end
        function prad_save(obj, h5filename)
            % Saves Simple Radiograph object to HDF5 file
            obj.validate();
            
            % Delete any existing file
            if (exist(h5filename, 'file') == 2)
                delete(h5filename);
                if (exist(h5filename, 'file') == 2)
                    error(['The existing HDF5 file: "' h5filename '" could not be overwritten.' newline 'Perhaps you do not have permissions to delete this file.' newline 'For example, perhaps the file is open and being read by another program, like HDFView? If so, close the file in that program.' newline 'Or, perhaps this file handle is currently open in MATLAB? Try the command fclose(''all'') to close all current file handles.' newline 'If all else fails, try manually deleting the file in question, or change your desired output filename to something else.' newline 'Once you''re done troubleshooting, run your script again.']);
                end
            end
            
            % Create new file and save object into it
            save_prad_group(obj, h5filename, '/')
        end
    end
end

File: Python/pradformat/simple_inversion.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
simple_inversion.py: Read and write the HDF5 pradformat "Simple Inversion" file structure

Created by Scott Feister on Mar 2 2021
"""

import h5py
import numpy as np
from datetime import datetime
from ._h5sanitize import _h5sanitize
from .__version__ import __version__

PRADFORMAT_VERSION = __version__ # awkward work-around to get __version__ variable into class

class SimpleInversion(object):
    """
    Object for storing data and attributes of an inverted SimpleRadiograph.
    """
    object_type = "inversion"
    inversion_type = "simple"

    # Categorize the above/below public properties as required or optional
    __req_ds = [] # Required datasets
    __opt_ds = ["phi", "defl_ax1", "defl_ax2"] # Optional datasets
    __req_atts = [  # Required attributes
        "object_type", 
        "inversion_type", 
        "pradformat_version", 
        "dr", 
        "source_object_dist", 
        "object_image_dist", 
        "spec_name", 
        "spec_mass", 
        "spec_charge", 
        "spec_energy", 
        ]
    __opt_atts = [  # Optional attributes
        "dr_ax2", 
        "source_radius", 
        "label", 
        "description", 
        "file_date", 
        "radiograph_filename", 
        "fields_filename", 
        ]

    def __init__(self, h5filename=None, template=None):
        self.phi = None
        self.defl_ax1 = None
        self.defl_ax2 = None
        
        self.pradformat_version = PRADFORMAT_VERSION
        self.dr = None
        self.source_object_dist = None
        self.object_image_dist = None
        self.spec_name = None
        self.spec_mass = None
        self.spec_charge = None
        self.spec_energy = None
        self.dr_ax2 = None
        self.source_radius = None
        self.label = None
        self.description = None
        self.file_date = datetime.now().strftime("%Y-%m-%d")
        self.radiograph_filename = None
        self.fields_filename = None
        
        if not isinstance(template, type(None)):
            # Set any matching properties in the template object to properties in this object
            all_props = set(self.__req_ds) | set(self.__opt_ds) | set(self.__req_atts) | set(self.__opt_atts) # all properties
            sticky_props = set(["inversion_type", "object_type", "pradformat_version", "file_date"]) # Don't change any of these properties to match the template object values
            for prop in (all_props - sticky_props):
                try:
                    setattr(self, prop, getattr(template, prop))
                except:
                    pass
        
        if not isinstance(h5filename, type(None)):
            # Load in all properties present in the HDF5 file
            self.load(h5filename)

    def __str__(self):
        mystr = "Simple Inversion Object\n\n"
        mystr += "Required datasets: \n"
        for ds in self.__req_ds:
            data = getattr(self, ds)
            mystr += "  " + ds + ": "
            if isinstance(data, np.ndarray):
                mystr += str(data.shape) + " ndarray, dtype " + str(data.dtype)
            else:
                mystr += str(data)
            mystr += "\n"
                
        mystr += "\nOptional datasets: \n"
        for ds in self.__opt_ds:
            data = getattr(self, ds)
            mystr += "  " + ds + ": "
            if isinstance(data, np.ndarray):
                mystr += str(data.shape) + " ndarray, dtype " + str(data.dtype)
            else:
                mystr += str(data)
            mystr += "\n"

        mystr += "\nRequired attributes: \n"
        for att in self.__req_atts:
            mystr += "  " + att + ": " + str(getattr(self, att)) + "\n"
            
        mystr += "\nOptional attributes: \n"
        for att in self.__opt_atts:
            mystr += "  " + att + ": " + str(getattr(self, att)) + "\n"
        
        return mystr

    def validate(self):
        # Validate that all required properties have been set.
        # If some are not, throw an error.
        for ds in self.__req_ds:
            if isinstance(getattr(self, ds), type(None)):
                raise Exception('Please assign a value to all required properties.\n The following required dataset property is not yet assigned: {0}.\n Assign a value via "object.{0} = value" and try again.'.format(ds))
        for att in self.__req_atts:
            if isinstance(getattr(self, att), type(None)):
                raise Exception('Please assign a value to all required properties.\n The following required attribute property is not yet assigned: {0}.\n Assign a value via "object.{0} = value" and try again.'.format(att))
        return

    def save(self, filename, compression_opts=4):
        """ Saves SimpleInversion object to HDF5 file"""
        self.validate() # Before opening a new HDF5 file, check that all required properties are set
        
        # Create file, overwriting if file already exists
        with h5py.File(filename, "w") as f: 
            # Write required datasets to file
            for ds in self.__req_ds:
                data = getattr(self, ds)
                if not np.isscalar(data):
                    f.create_dataset(ds, data=data, compression="gzip", compression_opts=compression_opts)
                else:
                    f.create_dataset(ds, data=data) # Don't compress scalar datasets (e.g. single-element arrays)
                    
            # Write optional datasets to file
            for ds in self.__opt_ds:
                data = getattr(self, ds)
                if not isinstance(data, type(None)):
                    if not np.isscalar(data):
                        f.create_dataset(ds, data=data, compression="gzip", compression_opts=compression_opts) # Compress only for datasets of length greater than one
                    else:
                        f.create_dataset(ds, data=data)
                
            # Write required attributes to file
            for att in self.__req_atts:
                f.attrs[att] = getattr(self, att)
                
            # Write optional attributes to file
            for att in self.__opt_atts:
                if not isinstance(getattr(self, att), type(None)): 
                    f.attrs[att] = getattr(self, att)
    
    def load(self, h5filename):
        """ Load data from HDF5 file into this object """
        with h5py.File(h5filename, "r") as f: 
            # Read in required datasets
            for ds in self.__req_ds:
                setattr(self, ds, f["/" + ds][()])
                
            # Read in optional datasets
            for ds in self.__opt_ds:
                if ds in f.keys():
                    setattr(self, ds, f["/" + ds][()])
                
            # Read in required attributes
            for att in self.__req_atts:
                if att not in ["object_type", "inversion_type"]:
                    setattr(self, att, _h5sanitize(f.attrs[att]))
                
            # Read in optional attributes
            for att in self.__opt_atts:
                if att in f.attrs.keys(): 
                    setattr(self, att, _h5sanitize(f.attrs[att]))
                   
if __name__ == "__main__":
    pass
